bitbots_vision/bitbots_vision/vision_modules/yoeo_handler.py

```diff
@@ -0,0 +1,207 @@
+        self._segmentations[class_name] = moveaxis(segmentation, 0, -1)
+
+    def set_image(self, img) -> None:
+        self._reset()
+        self._update_image(img)
+
+    def _reset(self) -> None:
+        self._det_candidates = defaultdict(list)
+        self._segmentation = None
+        self._segmentations = defaultdict(list)
+
+    def _update_image(self, img) -> None:
+        self._image = img
+
+    @abstractmethod
+    def _compute_new_prediction(self) -> None:
+        ...
+
+
+class YOEOHandlerPytorch(YOEOHandlerTemplate):
+    """
+    Using Pytorch to get YOEO predictions
+    """
+
+    def __init__(self, config, model_path):
+        super().__init__(config, model_path)
+
+        weights_path = join(model_path, "weights", "yoeo.pth")
+        config_path = join(model_path, "config", "yoeo-rev-7.cfg")
+
+        self._model = torch_models.load_model(config_path, weights_path)
+
+    def _compute_new_prediction(self) -> None:
+        detections, segmentation = torch_detect.detect_image(self._model,
+                                                             cv2.cvtColor(self._image, cv2.COLOR_BGR2RGB),
+                                                             conf_thres=self._det_confidence_thresh,
+                                                             nms_thres=self._iou_non_max_suppression_thresh)
+
+        for detection in detections:
+            c = Candidate.from_x1y1x2y2(*detection[0:4].astype(int), detection[4].astype(float))
+            self._det_candidates[self._det_class_names[int(detection[5])]].append(c)
+
+        self._segmentation = segmentation
+
+
+class YOEODetector(CandidateFinder):
+    def __init__(self, yoeo_handler: YOEOHandlerTemplate):
+        super().__init__()
+
+        self._yoeo_handler = yoeo_handler
+
+    def set_image(self, image) -> None:
+        self._yoeo_handler.set_image(image)
+
+    @abstractmethod
+    def get_candidates(self):
+        ...
+
+    def compute(self):
+        self._yoeo_handler.predict()
+
+
+class YOEOBallDetector(YOEODetector):
+    def __init__(self, yoeo_handler: YOEOHandlerTemplate):
+        super().__init__(yoeo_handler)
+
+    def get_candidates(self) -> List[Candidate]:
+        return self._yoeo_handler.get_detection_candidates_for("ball")
+
+
+class YOEOGoalpostDetector(YOEODetector):
+    def __init__(self, yoeo_handler: YOEOHandlerTemplate):
+        super().__init__(yoeo_handler)
+
+    def get_candidates(self):
+        return self._yoeo_handler.get_detection_candidates_for("goalpost")
+
+
+class YOEORobotDetector(YOEODetector):
+    def __init__(self, yoeo_handler: YOEOHandlerTemplate):
+        super().__init__(yoeo_handler)
+
+    def get_candidates(self):
+        return self._yoeo_handler.get_detection_candidates_for("robot")
+
+
+class YOEOSegmentationTemplate(ABC):
+    def __init__(self, yoeo_handler: YOEOHandlerTemplate):
+        self._yoeo_handler = yoeo_handler
+
+    @abstractmethod
+    def get_segmentation(self):
```

Comment:
In other parts of the vision `get_mask()` is used to describe this behavior. Especially in the field boundary detector. Maybe we need to fiddle a bit around with field boundary and line detectors too as we need to introduce common interfaces for lines and the field boundary mask. Currently, the field boundary is described by either a polygon line or a mask drawn by it and as YOEO is only producing masks it might result in issues.